File: sdks/csharp/src/NovaBilling/Types/ChargeFilterDto.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using NovaBilling.Core;

namespace NovaBilling;

[Serializable]
public record ChargeFilterDto : IJsonOnDeserialized
{
    [JsonExtensionData]
    private readonly IDictionary<string, JsonElement> _extensionData =
        new Dictionary<string, JsonElement>();

    /// <summary>
    /// Filter key (must match metric filter)
    /// </summary>
    [JsonPropertyName("key")]
    public required string Key { get; set; }

    /// <summary>
    /// Subset of allowed values
    /// </summary>
    [JsonPropertyName("values")]
    public IEnumerable<string> Values { get; set; } = new List<string>();

    /// <summary>
    /// Override properties for this filter
    /// </summary>
    [JsonPropertyName("properties")]
    public Dictionary<string, object?>? Properties { get; set; }

    [JsonIgnore]
    public ReadOnlyAdditionalProperties AdditionalProperties { get; private set; } = new();

    void IJsonOnDeserialized.OnDeserialized() =>
        AdditionalProperties.CopyFromExtensionData(_extensionData);

    /// <inheritdoc />
    public override string ToString()
    {
        return JsonUtils.Serialize(this);
    }
}

File: sdks/php/src/Types/ChargeFilterDto.php
<?php

namespace NovaBilling\Types;

use NovaBilling\Core\Json\JsonSerializableType;
use NovaBilling\Core\Json\JsonProperty;
use NovaBilling\Core\Types\ArrayType;

class ChargeFilterDto extends JsonSerializableType
{
    /**
     * @var string $key Filter key (must match metric filter)
     */
    #[JsonProperty('key')]
    public string $key;

    /**
     * @var array<string> $values Subset of allowed values
     */
    #[JsonProperty('values'), ArrayType(['string'])]
    public array $values;

    /**
     * @var ?array<string, mixed> $properties Override properties for this filter
     */
    #[JsonProperty('properties'), ArrayType(['string' => 'mixed'])]
    public ?array $properties;

    /**
     * @param array{
     *   key: string,
     *   values: array<string>,
     *   properties?: ?array<string, mixed>,
     * } $values
     */
    public function __construct(
        array $values,
    ) {
        $this->key = $values['key'];
        $this->values = $values['values'];
        $this->properties = $values['properties'] ?? null;
    }

    /**
     * @return string
     */
    public function __toString(): string
    {
        return $this->toJson();
    }
}
